### ISSUE TITLE
Keyboard focus does not pass.

### ISSUE THREAD

**Issue by apodtele** [mintty/mintty#1206]
Launching [ftgamma](https://github.com/ubawurinna/freetype-windows-binaries/blob/master/demos/win64/ftgamma.exe) from mintty opens its window in the background without passing the keyboard focus. Cmd, on the contrary, does it properly and passes the focus. This is observed with Git-for-Windows 2.40.0 based on MSYS2. Ftgamma is is simple Windows program with nothing beyond basic CreateWindow and GDI.

**Comment by apodtele** [comment]
Windows 11 behaves just weirdly when I start standard _pbrush.exe_ from mintty: pbrush.exe starts in the background but mintty also loses focus. Wow.

**Comment by mintty** [comment]
Where is the source of this ftgamma and can it be compiled with cygwin tools?
Does it happen on Windows 10 with this demo tool?

**Comment by mintty** [comment]
> Launching ftgamma from mintty ...

I suppose you mean running the program from a command line.
This is not really launching from mintty. It is the shell that launches the programm, mintty isn't involved.
I am therefore inclined to close the issue as invalid. It must be some weird interworking triggered by the launched program in Windows 11. Leaving open yet for a while as someone may come up with an analysis.

**Comment by apodtele** [comment]
Paint Brush (pbrush.exe) comes standard with any Windows version since Windows 95. If you can start that from a mintty command line without any focus surprises on Windows 11 then so be it. Close this as "invalid" or "works for me", or help me find where to file this bug report..

**Comment by mintty** [comment]
There is no pbrush.exe in my Windows 10 installation.
What about mspaint?
For clarity please answer my question: how exactly do you "launch" the program?

**Comment by apodtele** [comment]
Yes pbrush and mspaint are probably the same.  I guess Windows 11 resurrected the old name.

**Comment by mintty** [comment]
It seems to be a problem only on Windows 11 which I don't have so I cannot analyse it.

**Comment by apodtele** [comment]
Perhaps this: https://betanews.com/2023/01/05/windows-11-22h2-has-introduced-an-irritating-focus-bug-in-file-explorer/

**Comment by mintty** [comment]
Are you on Windows "fast release track"? Can you downgrade and retest? Or just wait until the fix...
To check whether a workaround is possible, you could use cygwin or MSYS2 (with packages gcc and make), recompile mintty with `make CCOPT=-Ddebug_messages`, cut or mark the trace where the launch command is entered, and upload it here.

**Comment by apodtele** [comment]
This is regular Windows 11 22H2 on the brand new Dell. The diagnostic shows that _explorer.exe_ interjects twice and steals the focus. Besides _mspaint.exe/pbrush.exe_, _calc.exe_ reliably starts out of focus for me. The calculator taskbar icon blinks too to indicate that it could not get focus.. Weird... I uninstalled McAfee, which did not help, but left a bunch of Dell software on for now.

**Comment by apodtele** [comment]
> Under `mintty` should use `cygstart` to launch _Windows_ GUIs e.g. `$ cygstart pbrush`

Please educate me why this is necessary. Or, why cannot `bash` do it?

Also, `cygstart` does not come with git-for-windows. Which package is it a part of?

**Comment by mintty** [comment]
> The diagnostic shows that explorer.exe interjects twice and steals the focus.

Which diagnostic actually? Is there any unusual message sending towards mintty that could be handled as a workaround?
Mintty message diagnostic can be enabled with compiling `make CCOPT=-Ddebug_messages`. If someone can demonstrate a difference between flawed behaviour (Windows 11?) and normal behaviour (Windows 10?) maybe we can get further.

**Comment by apodtele** [comment]
[Video](https://github.com/git-for-windows/git/issues/4360#issuecomment-1661282120) posted.

**Comment by thedmd** [comment]
I have this issue on Windows 11 22H2 (OS Build 22621.1992), on Windows 10 it is working fine.
Windows 11 was installed from ISO, it was not an upgrade from 10.

**Comment by apodtele** [comment]
Thanks, I am glad that I am not alone. 

I don't know how _calc_ or _mspaint_ designed. My programs CreateWindow and ShowWindow from a child thread expecting the active keyboard focus there. This may or may not be an important reason why the main thread started by _mintty_ mistakenly keeps the focus.

**Comment by mintty** [comment]
I have now tested this with Windows 11 and I do not reproduce the issue.
Running calc from a shell in mintty transfers focus to calc for me.

**Comment by apodtele** [comment]
I am not sure if it matters, but I guess that you testes a current msys2 installation with bash. I would like to try your setup before squarely blaming Git-for-Windows.

**Comment by mintty** [comment]
I run cygwin. I can hardly imagine this should make a difference but who knows.
Please also check your $PS1 prompt, or clear it (PS1=.) for a cross-test.

**Comment by thedmd** [comment]
Issue still persist with:
![image](https://github.com/mintty/mintty/assets/1197433/52e6a566-178d-47ae-b5f0-e64b499f61a2)
I would put my money on bug in Explorer, on Windows 10 focus still behave ok.

**Comment by mintty** [comment]
Current version is actually 3.4.8.

> Please also check your $PS1 prompt, or clear it (PS1=.) for a cross-test.

Just to make sure there is no escape sequence that would explicitly trigger this behaviour.

**Comment by thedmd** [comment]
By $PS1 do you mean to use Windows PowerShell or PowerShell passed through mintty? Sorry, I'm not well versed in terminals.

So far I'm testing using git from Windows PowerShell and here focus behave correctly.

**Comment by mintty** [comment]
No, PS1 is a Unix/Linux/Cygwin shell variable to configure the prompt string. No PowerShell involved.
Please set `PS1=.` then run calc again.

**Comment by thedmd** [comment]
I did set `PS1=.` and typed `calc`. Calculator showed up under terminal window.

This is still one from latest Git for Windows. I need to look into installing latest version independently.

<img width="497" alt="image" src="https://github.com/mintty/mintty/assets/1197433/b16647a9-700a-454e-baf3-836f47a7d96d">

**Comment by thedmd** [comment]
I installed latest nightly of msys2. Same result.

<img width="733" alt="image" src="https://github.com/mintty/mintty/assets/1197433/33b02f4d-fa00-47b1-8331-e99d3b0354d7">

**Comment by apodtele** [comment]
I filed with msys2-runtime as a point of departure from cygwin.

**Comment by mintty** [comment]
Did you cross-check on Windows 11/cygwin?
I plan to cross-check with Msys but my Windows 11 machine is currently packed away...

**Comment by apodtele** [comment]
Yes, I can confirm that _cygwin_ passes focus. _msys2_ does not.

**Comment by mintty** [comment]
I do not reproduce that. Just installed MSYS2 and Git for Windows on Windows 11 and it works for me.

**Comment by thedmd** [comment]
Version of windows I have:
```
Edition		Windows 11 Pro
Version		22H2
Installed on	‎11.‎04.‎2023
OS build	22621.2283
Experience	Windows Feature Experience Pack 1000.22662.1000.0
```

**Comment by apodtele** [comment]
```
Edition		Windows 11 Home
Version		22H2
Installed on	‎3/‎19/‎2023
OS build	22621.2283
Experience	Windows Feature Experience Pack 1000.22662.1000.0
```

**Comment by ElvenSpellmaker** [comment]
I just updated my Git for Windows (2.39.2 - mintty 3.6.3) and I can reproduce this too, never on Cygwin. So it seems to be a bug with msys2.

```
$ cmd /c ver

Microsoft Windows [Version 10.0.22621.2861]
```

(I'm _actually_ on 22H3 despite that the output above might hint at me being 22H2. I downloaded the package and installed manually).

**Comment by apodtele** [comment]
Other issues related to focus get some attention lately #1207, #1242, #1249.  Perhaps it will all resolve one day.

**Comment by mintty** [comment]
Those other issues are all related to multiple tabs, so I guess it's something different.
Can someone test whether it happens on plain MSYS2 (not Git for W)?

**Comment by apodtele** [comment]
Confirmed on vanilla msys2/msys2-runtime/issues/173.

This bug was introduced then the tab focus work begun. It may be a coincidence.

**Comment by apodtele** [comment]
Switched to Windows Terminal with bash (Git-for-Windows) because of this issue.

**Comment by mintty** [comment]
Just to confirm, after some updates I can meanwhile also reproduce this issue in MSYS2 (or Git) on Windows 11.

**Comment by chipbite** [comment]
...and I tested a bit more to reproduce the difference. I don't know why, and if this is reproducing the issue or not - I am very unsure. BUT, this behavior is consistent on my machine:

```bash
$ echo -e '#!'"/bin/sh\nexec notepad.exe &\n" > testworks.sh; chmod +x testworks.sh
$ ./testworks.sh
$ echo -e "exec notepad.exe &\n" > testfails.sh; chmod +x testfails.sh
$ ./testfails.sh
$
```

Where `testworks` will launch the ui app with focus as expected, and `testfails` at the back/bottom of the window stack instead. Not sure if this helps, but.

**Comment by apodtele** [comment]
Do bash or dash ever call SetFocus? Probably not, but there could be a race for focus between the terminal and the launched app. The terminal should never be in this race, not just lose sometimes.

**Comment by apodtele** [comment]
I very much doubt that _bash_ or _dash_ care about what programs do beyond their return status. They probably do not care if they spin threads which open Windows of if those windows get focus from the window manager. 

It is a normal behavior that a newly created window gets the keyboard focus. The program does not even need to SetFocus or be polite by waiting. The last window gets the focus, unless some other window decides to SetFocus. mintty does SetFocus occasionally, does it not?

**Comment by mintty** [comment]
There might be some subtle interworking that differs among shells, e.g. around prompt display.
My next test cases (no access to Windows 11 right now) would be `PS1=.` and `(calc; sleep 1)`.
I'll check about invocations of SetFocus, SetActiveWindow or ShowWindow.
